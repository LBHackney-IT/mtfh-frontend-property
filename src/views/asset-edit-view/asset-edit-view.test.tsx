import React from "react";

import { getAssetV1, mockAssetV1, render, server } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";

import { AssetEditView } from ".";

const assetData = {
  id: "15adc44b-6fde-46e8-af9c-e18b1495c9ab",
  assetId: "100021045676",
  assetType: "LettableNonDwelling",
  rootAsset: null,
  parentAssetIds: null,
  isActive: false,
  assetLocation: null,
  assetAddress: {
    uprn: "100021045676",
    addressLine1: "51 GREENWOOD ROAD - FLAT B",
    addressLine2: "",
    addressLine3: "",
    addressLine4: "",
    postCode: "E8 1NT",
    postPreamble: null,
  },
  assetManagement: {
    agent: null,
    areaOfficeName: null,
    isCouncilProperty: false,
    managingOrganisation: null,
    managingOrganisationId: "00000000-0000-0000-0000-000000000000",
    owner: null,
    isTMOManaged: false,
    propertyOccupiedStatus: null,
    propertyOccupiedStatusReason: null,
    isNoRepairsMaintenance: false,
    councilTaxType: null,
    councilTaxLiability: null,
    isTemporaryAccomodation: true,
    readyToLetDate: false,
  },
  assetCharacteristics: null,
  tenure: null,
  versionNumber: 18,
  patches: [
    {
      id: "bd0a8e2b-c3b5-4628-aa33-8e7509d5eac6",
      parentId: "8d4fb05d-3ff5-48b7-a17a-71fcb27a66a8",
      name: "SN4",
      patchType: "patch",
      domain: "MMH",
      responsibleEntities: [],
      versionNumber: null,
    },
  ],
};

const llpgAddressData = {
  data: {
    address: [
      {
        line1: "FLAT B",
        line2: "51 GREENWOOD ROAD",
        line3: "HACKNEY",
        line4: "",
        town: "LONDON",
        postcode: "E8 1NT",
        UPRN: 100021045676,
      },
    ],
    page_count: 1,
    total_count: 1,
  },
  statusCode: 200,
};

beforeEach(() => {
  jest.resetAllMocks();

  server.use(
    rest.get("/api/v1/assets/:id", (req, res, ctx) => {
      return res(ctx.json(assetData));
    }),
  );

  server.use(
    rest.get("/api/v1/addresses", (req, res, ctx) => {
      return res(ctx.json(llpgAddressData));
    }),
  );

  server.use(
    rest.patch(`/api/v1/assets/${assetData.id}/address`, (req, res, ctx) => {
      return res(ctx.status(204));
    }),
  );
});

test("renders the error on asset load failure", async () => {
  server.use(getAssetV1(undefined, 500));
  render(<AssetEditView />, {
    url: `/property/edit/${mockAssetV1.id}`,
    path: "/property/edit/:assetId",
  });

  await screen.findByRole("alert");
});

test("renders the whole 'Edit property address' view", async () => {
  const { container } = render(<AssetEditView />, {
    url: `/property/edit/${assetData.id}`,
    path: "/property/edit/:assetId",
  });

  // This allows the test to wait for the page to be populated, after receiving data from the mock Address and Asset APIs
  await waitFor(() => expect(screen.getAllByRole("heading")).toHaveLength(3));

  expect(container).toMatchSnapshot();
});

test("the current address from the asset is updated using the LLPG address suggestion", async () => {
  render(<AssetEditView />, {
    url: `/property/edit/${assetData.id}`,
    path: "/property/edit/:assetId",
  });

  // This allows the test to wait for the page to be populated, after receiving data from the mock Address and Asset APIs
  await waitFor(() => expect(screen.getAllByRole("heading")).toHaveLength(3));

  // Assert LLPG Address Line 1 is FLAT B
  const llpgAddressLine1 = screen.getByTestId("addressLine1");
  expect(llpgAddressLine1).toHaveValue("FLAT B");

  // Assert Asset Address Line 1 is 51 GREENWOOD ROAD - FLAT B
  const assetAddressLine1 = screen.getByTestId("asset-address-line-one");
  expect(assetAddressLine1).toHaveValue("51 GREENWOOD ROAD - FLAT B");

  // Assert LLPG AddrLine2 is 51 GREENWOOD ROAD
  const llpgAddressLine2 = screen.getByTestId("addressLine2");
  expect(llpgAddressLine2).toHaveValue("51 GREENWOOD ROAD");

  // Assert Asset AddrLine2 is ""
  const assetAddressLine2 = screen.getByTestId("asset-address-line-two");
  expect(assetAddressLine2).toHaveValue("");

  // Assert LLPG AddrLine3 is HACKNEY
  const llpgAddressLine3 = screen.getByTestId("addressLine3");
  expect(llpgAddressLine3).toHaveValue("HACKNEY");

  // Assert Asset AddrLine3 is ""
  const assetAddressLine3 = screen.getByTestId("asset-address-line-three");
  expect(assetAddressLine3).toHaveValue("");

  // Assert LLPG AddrLine4 is LONDON
  const llpgAddressLine4 = screen.getByTestId("addressLine4");
  expect(llpgAddressLine4).toHaveValue("LONDON");

  // Assert Asset AddrLine4 is ""
  const assetAddressLine4 = screen.getByTestId("asset-address-line-four");
  expect(assetAddressLine4).toHaveValue("");

  // User click on "Update to this address"
  const updateButton = screen.getByRole("button", { name: "Update to this address" });

  // The await is required, as it allows the PATCH API call to be intercepted and the Current Asset Address value to be replaced with the LLPG suggestion.
  // Without it, the test may pass but the below expects would not work as expected.
  await waitFor(() => {
    userEvent.click(updateButton);

    // Assert Asset Address Line 1 is FLAT B, the same as LLPG Address Line 1
    expect(assetAddressLine1).toHaveValue("FLAT B");

    // Assert Asset AddrLine2 is 51 GREENWOOD ROAD, the same as LLPG Address Line 2
    expect(assetAddressLine2).toHaveValue("51 GREENWOOD ROAD");

    // Assert Asset AddrLine3 is HACKNEY, the same as LLPG Address Line 3
    expect(assetAddressLine3).toHaveValue("HACKNEY");

    // Assert Asset AddrLine4 is LONDON, the same as LLPG Address Line 1
    expect(assetAddressLine4).toHaveValue("LONDON");
  });
});

test("form action buttons are rendered and are enabled", async () => {
  render(<AssetEditView />, {
    url: `/property/edit/${assetData.id}`,
    path: "/property/edit/:assetId",
  });

  // This allows the test to wait for the page to be populated, after receiving data from the mock Address and Asset APIs
  await waitFor(() => expect(screen.getAllByRole("heading")).toHaveLength(3));

  // Assert "Update to this button" is in the DOM and is enabled
  const updateButton = screen.getByRole("button", { name: "Update to this address" });
  const cancelButton = screen.getByText("Cancel edit address");

  expect(updateButton).toBeInTheDocument();
  expect(updateButton).toBeEnabled();

  expect(cancelButton).toBeInTheDocument();
  expect(cancelButton).toBeEnabled();
});

test("the URL of the page includes property/edit followed by the asset GUID", async () => {
  render(<AssetEditView />, {
    url: `/property/edit/${assetData.id}`,
    path: "/property/edit/:assetId",
  });

  expect(window.location.pathname).toContain(`/property/edit/${assetData.id}`);
});

test("it shows the 'Back to asset' link", async () => {
  render(<AssetEditView />, {
    url: `/property/edit/${assetData.id}`,
    path: "/property/edit/:assetId",
  });

  const backLink = await screen.findByText("Back to asset");
  expect(backLink).toBeVisible();
});
