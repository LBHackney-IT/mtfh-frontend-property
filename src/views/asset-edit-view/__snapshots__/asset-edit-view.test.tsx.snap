// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the whole 'Edit property address' view 1`] = `
<div>
  <a
    class="govuk-back-link lbh-back-link lbh-link--no-visited-state"
    href="/property/15adc44b-6fde-46e8-af9c-e18b1495c9ab"
  >
    Back to asset
  </a>
  <h1
    class="lbh-heading-h1"
  >
    Edit property address
  </h1>
  <span
    class="govuk-caption-m lbh-caption"
  >
    New Addresses are suggested from the Local Gazetteer to bring some standardisation.
  </span>
  <div
    class="mtfh-address-details"
  >
    <section>
      <div
        id="edit-address-form"
      >
        <form
          action="#"
        >
          <h3
            class="lbh-heading-h3"
          >
            Suggestion from the Local Gazetteer
          </h3>
          <div
            class="govuk-form-group lbh-form-group"
          >
            <label
              class="govuk-label lbh-label"
              for="addressLine1"
            >
              Address line 1*
            </label>
            <span
              class="govuk-error-message lbh-error-message"
              id="addressLine1-input-error"
            >
              <span
                class="govuk-visually-hidden"
              >
                Error:
              </span>
               
            </span>
            <input
              class="govuk-input lbh-input"
              data-testid="addressLine1"
              id="addressLine1"
              name="addressLine1"
              type="text"
              value="FLAT B"
            />
          </div>
          <label
            class="govuk-label lbh-label"
            for="addressLine2"
          >
            Address line 2
          </label>
          <input
            class="govuk-input lbh-input"
            data-testid="addressLine2"
            id="addressLine2"
            name="addressLine2"
            type="text"
            value="51 GREENWOOD ROAD"
          />
          <label
            class="govuk-label lbh-label"
            for="addressLine3"
          >
            Address line 3
          </label>
          <input
            class="govuk-input lbh-input"
            data-testid="addressLine3"
            id="addressLine3"
            name="addressLine3"
            type="text"
            value="HACKNEY"
          />
          <label
            class="govuk-label lbh-label"
            for="addressLine4"
          >
            Address line 4
          </label>
          <input
            class="govuk-input lbh-input"
            data-testid="addressLine4"
            id="addressLine4"
            name="addressLine4"
            type="text"
            value="LONDON"
          />
          <div
            class="govuk-form-group lbh-form-group"
          >
            <label
              class="govuk-label lbh-label"
              for="postcode"
            >
              Postcode
            </label>
            <span
              class="govuk-error-message lbh-error-message"
              id="Postcode-input-error"
            >
              <span
                class="govuk-visually-hidden"
              >
                Error:
              </span>
               
            </span>
            <input
              class="govuk-input lbh-input"
              data-testid="postcode"
              id="postcode"
              name="postcode"
              type="text"
              value="E8 1NT"
            />
          </div>
          <div
            class="form-actions"
          >
            <button
              class="govuk-button lbh-button"
              data-module="govuk-button"
              id="submit-address-button"
              type="submit"
            >
              Update to this address
            </button>
            <a
              class="govuk-button govuk-secondary lbh-button lbh-button--secondary"
              href="/property/15adc44b-6fde-46e8-af9c-e18b1495c9ab"
            >
              Cancel edit address
            </a>
          </div>
        </form>
      </div>
    </section>
    <section>
      <h3
        class="lbh-heading-h3"
      >
        Current address
      </h3>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="asset-address-line-one-input"
        >
          Address line 1*
        </label>
        <input
          class="govuk-input lbh-input"
          data-testid="asset-address-line-one"
          disabled=""
          id="asset-address-line-one-input"
          name="asset-address-line-one"
          required=""
          type="text"
          value="51 GREENWOOD ROAD - FLAT B"
        />
      </div>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="asset-address-line-two-input"
        >
          Address line 2
        </label>
        <input
          class="govuk-input lbh-input"
          data-testid="asset-address-line-two"
          disabled=""
          id="asset-address-line-two-input"
          name="asset-address-line-two"
          type="text"
          value=""
        />
      </div>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="asset-address-line-three"
        >
          Address line 3
        </label>
        <input
          class="govuk-input lbh-input"
          data-testid="asset-address-line-three"
          disabled=""
          id="asset-address-line-three-input"
          name="asset-address-line-three"
          type="text"
          value=""
        />
      </div>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="asset-address-line-four-input"
        >
          Address line 4
        </label>
        <input
          class="govuk-input lbh-input"
          data-testid="asset-address-line-four"
          disabled=""
          id="asset-address-line-four-input"
          name="asset-address-line-four"
          type="text"
          value=""
        />
      </div>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="asset-postcode-input"
        >
          Postcode
        </label>
        <input
          class="govuk-input lbh-input"
          data-testid="asset-postcode"
          disabled=""
          id="asset-postcode-input"
          name="asset-postcode"
          required=""
          type="text"
          value="E8 1NT"
        />
      </div>
    </section>
  </div>
</div>
`;
